### ISSUE TITLE
 accurate QPS from user's perspective

### ISSUE THREAD

**Issue by panffeng** [elastic/elasticsearch#27982]
**Describe the feature**: Elasticsearch keeps the stats for indices and shards as _stat shows. From user's perspective, users are often at loss to figure out query rates of searching and indexing for cluster level or among clusters, which should be ideally consistent to the QPS user sends queries with.  Yet, the numbers from _stat are affected by many factors like primary shard numbers, replica numbers and different settings for various indices.

https://discuss.elastic.co/t/marvel-what-does-the-search-rate-mean/35260 shows a frequent confusion users are dealing with when _stat is referred directly.

The user's perspective stats should be kept about total searching queries executed by coordinate nodes for searching QPS and total indexing queries done by primary shards for indexing QPS.

The user's perspective stats are very important especially when there is more than one cluster serving users. The total of traffic distributed among clusters should be equal to the user's QPS.

So, it would be nice to have user's perspective stats at cluster level that give out user's indexing QPS or count and searching QPS or count directly.

**Comment by DaveCTurner** [comment]
The numbers given by the `_stats` API are important, and reflect the distributed nature of an Elasticsearch cluster. A single search request from the user can result in queries against a large number of shards in the cluster, and this amplification factor is captured in the stats currently reported. Similarly, an indexing request from the user can result in the document being indexed on multiple shards, as determined by the number of replicas for the index in question.

We have issues like #14576 and #22199 open to improve the documentation of these numbers to reduce any confusion they may generate, and warmly welcome contributions in this area.

Counting the number of requests as sent by the user effectively ignores this amplification and treats all requests as equal. I'm not sure that this is a meaningful number from Elasticsearch's point of view, and calculating it accurately is a pretty serious engineering challenge when you start to think about the failure modes of a distributed cluster.

I'll raising this for a quick discussion with the wider team when we return from the Christmas break to see if there are any other comments.

**Comment by panffeng** [comment]
Hi Dave,

thanks a lot for your reply.

_stats API are absolutely important for elasticsearch. They are essential to diagnosing, monitoring and operating elasticsearch clusters. The stats make perfect sense in the elasticsearch world.

Accurate QPS from user's perspective should be implemented as a separate API from _stats API.

The gap between elasticsearch stats and user's stats widens when there are several elasticsearch clusters in different colos.  Traffics are distributed among clusters. Users only report application level QPS of indexing queries and searching queries. From application level, the _stats API is very difficult to interpret if there are shard numbers, replica numbers and multiple indices involved. It would be also very difficult to find skewed traffic load and to allocate accordingly computing resources.

High availability of elasticsearch clusters covers most cases of node failures. Then the accurate QPS from user's perspective should be maintained by master nodes.

**Comment by DaveCTurner** [comment]
> It would be also very difficult to find skewed traffic load and to allocate accordingly computing resources.

From the point of view of Elasticsearch's resource needs, the numbers reported by the `_stats` API are what you should use to do this; an "external" request rate would not be the right thing to look at. From the point of view of your wider application it's possible that the "external" Elasticsearch request rate is currently a useful proxy for your resource needs, but even if that's true now I wouldn't expect it to be true as your application grows. I think you are looking for something like [APM](https://www.elastic.co/blog/elastic-apm-beta-released).

>  the accurate QPS from user's perspective should be maintained by master nodes.

To work correctly this'd need the master node to be involved in each request, which would become too much of a bottleneck at scale to be viable.

**Comment by panffeng** [comment]
The accurate QPS should be maintained by coordinates nodes, data nodes and the master nodes. Then periodically, master nodes get a sum from those distributive stats. In case of node failures, it is possible to lose some counts for a time slot. Master node will not participate each request.

When there are multiple clusters, a complex load-balancing strategy is applied. Typically queries are sent to the nearest clusters. As there is no direct way to ascertain the accurate QPS from user's perspective now, _stats API often complicates performance issues with traffic patterns and resource allocation.

**Comment by DaveCTurner** [comment]
We discussed this in Fix-it-Friday.

One suggestion was to use the [usage API](https://www.elastic.co/guide/en/elasticsearch/reference/6.1/cluster-nodes-usage.html) to count the number of times each node has serviced a search request. Since this is a per-node statistic, it does not suffer the consistency issues that we'd have to solve if we exposed this as a cluster-wide statistic instead.

It's clearly possible to calculate a cluster-wide QPS statistic using some kind of distributed counter as you describe, but it's complicated and lacks a strong enough justification for putting effort into it at the moment. In particular, your use case (load-balancing) should make its decisions in terms of the existing stats which represent the load on the clusters more accurately than a naive QPS statistic would do.
